feat!: Add data for new Legends ZA content

Madmadness65 opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-  [Bug]: If the PR is primarily a bug fix
-  [Move]: If a move has new or changed functionality
-  [Ability]: If an ability has new or changed functionality
-  [Item]: For new or modified items
-  [Mystery]: For new or modified Mystery Encounters
-  [Test]: If the PR is primarily adding or modifying tests
-  [UI/UX]: If the PR is changing UI/UX elements
-  [Audio]: If the PR is adding or changing music/sfx (may come with an associated assets PR)
-  [Sprite]: If the PR is adding or changing sprites (may come with an associated assets PR)
-  [Balance]: If the PR is related to game balance
-  [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-  [Dev]: If the PR is primarily changing something pertaining to development (lefthook hooks, linter rules, etc.)
-  [i18n]: If the PR is primarily adding/changing locale keys or key usage (may come with an associated locales PR)
-  [Docs]: If the PR is adding or modifying documentation (such as tsdocs/code comments)
-  [GitHub]: For changes to GitHub workflows/templates/etc
-  [Misc]: If no other category fits the PR
-->

<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small!)
-->

## What are the changes the user will see?

This adds data for the new mega evolutions, mega stones, new abilities, one new move, and a session migrator to accommodate the new mega stones. Additionally, Pokemon with forms that have unique categories will now properly display as such in the Pokedex.

## Why am I making these changes?
<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->
It is new official content that we want to be able to use.

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->
Adds new form data for all the new megas introduced in ZA and its DLC, all of the new mega stones (which had to bump all the form change items ID 50 and higher an additional 50 to fit in the new stones, complete with a session migrator), six new abilities with their full functionality with space reserved for two more we don't know about yet, and the one new move, Nihil Light, complete with it's functionality to ignore stats and hit Fairy types.

## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->
The new move Nihil Light in use:

https://github.com/user-attachments/assets/1eb914eb-123e-4b1a-a6ab-9347ddf43cf1

The Pokemon categories for alternate forms:

https://github.com/user-attachments/assets/e468638d-5adc-4eae-829c-4162de7acb94

## How to test the changes?
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->
Use the starter form overrides to test the new megas, or ability overrides to test the new abilities.

Nihil Light can be tested as-is as well, though moves learned on form change has not been merged yet.

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  ~~- [ ] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?~~
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  ~~- [ ] Have I made sure that any UI change works for both UI themes (default and legacy)?~~

Are there any localization additions or changes? If so:
- [x] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [x] If so, please leave a link to it here: https://github.com/pagefaultgames/pokerogue-locales/pull/252
- [x] Has the translation team been contacted for proofreading/translation?

Does this require any changes to the assets folder? If so:
  - [x] Has a PR been created on the [assets](https://github.com/pagefaultgames/pokerogue-assets) repo?
    - [x] If so, please leave a link to it here: https://github.com/pagefaultgames/pokerogue-assets/pull/8

Madmadness65 committed:
Add data for new Legends ZA content

This adds data for the new mega evolutions, mega stones, one new move, and a migration file to accommodate the new mega stones.
> [!WARNING]
> The version converter changes are currently commented out, ensure that they are uncommented when the PR is ready!

Madmadness65 committed:
Fix incorrect abilities for Mega Dragonite

Tough Claws was accidentally left as a placeholder.

Bertie690 commented on src/data/moves/move.ts:
```diff
@@ -5659,6 +5659,25 @@ export class NeutralDamageAgainstFlyingTypeMultiplierAttr extends VariableMoveTy
+   */
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
+    const multiplier = args[0] as NumberHolder;
+    if (target.isOfType(PokemonType.FAIRY)) {
```
For the record, this will not work as expected - TArrows' behavior is to _override_ the computed type multiplier.

A Dragon/fairy type will take 1x from Nihil Light and not 2x.

(To fix, change it to use the same base class as Freeze Dry which applies to individual type chart matchups)

(Incidentally, the entire reason i refactored these in #6241 is to make them consistent.)

DayKev replied:
We'll also want tests for the move.

Madmadness65 committed:
Base Nihil Light's attribute off of Freeze Dry

It will now correctly alter the type matchups instead of *always* being neutral effectiveness.

Madmadness65 committed:
Update Water Shuriken's edge case todo

It has been confirmed that Water Shuriken changes to be a 75 power basic attack when used by Mega Greninja.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Add Mega Z enum, add to relevant places

This does not yet add any additional data for the new Megas themselves, that will be filled out when they are all officially released with the DLC.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Add data for all new ZA DLC megas

Same as the base ZA megas, the form change method will remain commented out until sprites are complete for them.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Fix import error

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Add support for Pokemon forms with different categories

Madmadness65 committed:
Fix regular Emboar having wrong hidden ability

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Fix Tatsugiri generating wild Megas

It will once again only spawn as Curly, Droopy, or Stretchy forms.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Update Nihil Light attribute

Nihil Light's move attribute updated to better match the other `MoveTypeChartOverride` attributes.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Add new ZA mega abilities

Implementing Mega Sol required a small amount of refactoring weather-related code to accommodate the new `getEffectiveWeatherForUser` function. Existing functionality should not be impacted by this change.

Additionally, Meowstic's form change code for the Mega has been corrected to work properly (It is still commented out until the sprites are added).

Madmadness65 committed:
Add placeholder abilities for unknown abilities

The Mega Sol tests were failing with the missing ability ID enums, so placeholder entries have been added (with the benefit of not needing a version migration when they are properly added).

Madmadness65 committed:
Update Nihil Light stats to match Champions

PP was reduced from 10 to 5. Power was already previously lowered from 200 to 100.

damocleas committed:
add placeholder abilities for megas still not in Champions

Madmadness65 committed:
Special cased Magearna's and Tatsugiri's megas

Mega Magearna and Mega Tatsugiri's alternate forms should now appear visually unique in-game, but just display "Mega (Pokemon)" as the name without any extra prefixes or suffixes.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Remove no-longer-used battler tag

Madmadness65 committed:
Linting

Madmadness65 committed:
Fix unintentional Magearna forms in the wild

Only regular and original color will appear in the wild & in eggs as intended.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Fix test still using getAttackTypeMultiplier

Madmadness65 committed:
Update Mega Raichu abilities

X now has Electric Surge, and Y now has No Guard.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Update new species configs for new megas

The form change entries are now uncommented in preparation to be tested.

Co-Authored-By: Fabi <192151969+fabske0@users.noreply.github.com>

fabske0 committed:
add new mega form keys to species data registry

Madmadness65 committed:
Implement Eelevate and Fire Mane abilities

Additionally, ability distribution for Megas added in Champions' Regulation M-B has been updated.

Madmadness65 committed:
Merge branch 'beta' into za-content

Madmadness65 committed:
Update Eelevate to use updated stat boost code

Madmadness65 committed:
Linting

SirzBenjie commented on src/system/version-migration/versions/v1_12_0.ts:
```diff
@@ -0,0 +1,27 @@
+const shiftFormChangeItems: SessionSaveMigrator = {
+  version: "1.12.0",
+  migrate: (data: SessionSaveData) => {
+    // Shifting these up by 50 will work for now, but a more permanent solution will be desired in the future
```
The modifier branch uses high bit prefixes for item classes, so we won't run into this once 1.13 drops.

SirzBenjie commented on src/system/version-migration/versions/v1_12_0.ts:
```diff
@@ -0,0 +1,27 @@
+      }
+    }
+  },
+};
```
Please change this into a `for (const modifier of data.modifiers)` (unless there's a reason I'm not seeing for it to be like this).

Madmadness65 replied:
It was due to following code consistency in another migration file (specifically `1_7_0.ts`), which also migrated items.

DayKev replied:
That one needed to remove items from the `data.modifiers` array (it wasn't just changing values), so it needed the index number of the modifier in the array.

SirzBenjie commented on src/data/abilities/ab-attrs.ts:
```diff
@@ -948,6 +948,39 @@ export class PostDefendContactApplyStatusEffectAbAttr extends PostDefendAbAttr {
+ * Applies a status effect to the attacker when the Pokemon is hit by an attacking move.
+ * Unlike {@linkcode PostDefendContactApplyStatusEffectAbAttr}, this does not require the move to make contact.
+ */
+export class PostDefendApplyStatusEffectAbAttr extends PostDefendAbAttr {
```
Wondering if we should just reuse `postDefendConjtactApplyStatus`, drop the `Contact` part of the name, and add a parameter to the AbAttr class asking for contact. This seems like it repeats a good amount of the code otherwise.

Madmadness65 committed:
Update migration file

Madmadness65 committed:
Combine status effect application ability attributes

Since their function was very similar (one checked for contact while the other didn't), they have been combined and given a boolean to check if contact is required instead.

SirzBenjie commented on src/field/pokemon.ts:
```diff
@@ -2311,6 +2311,27 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+   * Otherwise, the current arena weather type is returned (or {@linkcode WeatherType.NONE}
+   * if weather is absent or suppressed).
+   */
+  public getEffectiveWeatherTypeForMoves(): WeatherType {
```
The method is only used in the test for mega sol. There's no other references to it.

I'm also not sure it makes sense for it to be on the `pokemon`...

SirzBenjie commented on src/ui/handlers/pokedex-page-ui-handler.ts:
```diff
@@ -2602,6 +2602,38 @@ export class PokedexPageUiHandler extends MessageUiHandler {
       // Setting the category
       if (isFormCaught) {
         this.pokemonCategoryText.setText(species.category);
+        switch (species.speciesId) {
+          case SpeciesId.FERALIGATR:
+          case SpeciesId.DARKRAI:
+          case SpeciesId.MEOWSTIC:
+          case SpeciesId.DRAMPA:
+            if (this.species.forms[this.formIndex].formKey === "mega") {
+              this.pokemonCategoryText.setText(
+                i18next.t(`pokemonCategory:mega${toTitleCase(SpeciesId[species.speciesId])}Category`),
+              );
+            }
+            break;
+          case SpeciesId.HOOPA:
+            if (this.formIndex === 1) {
+              this.pokemonCategoryText.setText(i18next.t("pokemonCategory:hoopaUnboundCategory"));
+            }
+            break;
+          case SpeciesId.CALYREX:
+            if (this.formIndex > 0) {
+              this.pokemonCategoryText.setText(i18next.t("pokemonCategory:calyrexRiderCategory"));
+            }
+            break;
+          case SpeciesId.PALAFIN:
+            if (this.formIndex === 1) {
+              this.pokemonCategoryText.setText(i18next.t("pokemonCategory:palafinHeroCategory"));
+            }
+            break;
+          case SpeciesId.GIMMIGHOUL:
+            if (this.formIndex === 1) {
+              this.pokemonCategoryText.setText(i18next.t("pokemonCategory:gimmighoulRoamingCategory"));
+            }
+            break;
+        }
```
This is not an ideal way of handling this. It ends up adding quite a bit of code and is a maintenance burden.

We should add the locale key for forms to the species data, if anything.

SirzBenjie commented on src/battle-scene.ts:
```diff
@@ -1868,8 +1866,11 @@ export class BattleScene extends SceneBase {
           return 0; // Don't give trainers Battle Bond Greninja, Froakie or Frogadier
         }
         return randSeedInt(2);
+      case SpeciesId.MAGEARNA:
```
Does this mean trainers can generate with mega magearna?

Madmadness65 replied:
No, this means regular Magearna and original color Magearna can appear in the wild/eggs/etc.

SirzBenjie committed:
fix: broken bulbapedia links for abilities

SirzBenjie committed:
tidy: deduplicate eelevate and beast boost stat calc

SirzBenjie commented on src/data/abilities/ab-attrs.ts:
```diff
@@ -918,22 +918,32 @@ export class PostDefendTerrainChangeAbAttr extends PostDefendAbAttr {
-      move.doesFlagEffectApply({ flag: MoveFlags.MAKES_CONTACT, user: attacker, target: pokemon })
+      (this.contactRequired
+        ? move.doesFlagEffectApply({ flag: MoveFlags.MAKES_CONTACT, user: attacker, target: pokemon })
+        : move.category !== MoveCategory.STATUS)
```
How will this interact with pollen puff on ally?

Madmadness65 replied:
Right now, if it targets an ally that has Spicy Spray, it will burn the user of Pollen Puff. I'm not sure if that's accurate with Champions, though.

Madmadness65 committed:
Remove duplicate getEffectiveWeather method

Was only duplicated and used for convenience in Mega Sol's tests. Now uses the function in `move.ts` instead.

github-actions[bot] committed:
Update assets submodule

github-actions[bot] committed:
Update assets submodule

SirzBenjie committed:
make mega sol remove defensive boost from snow/sandstorm

SirzBenjie committed:
Update tests to use new method

DayKev committed:
Merge branch 'beta' into za-content

DayKev committed:
Merge branch 'migrators' into za-content

DayKev committed:
Merge branch 'release' into za-content

fabske0 commented on src/enums/form-change-item.ts:
```diff
@@ -50,7 +50,54 @@ export enum FormChangeItem {
+  GOLISOPITE,
+  MAGEARNITE,
+  ZERAORITE,
+  SCOVILLAINITE,
+  GLIMMORANITE,
+  TATSUGIRINITE,
+  BAXCALIBRITE,
+
+  BLUE_ORB = 100,
   RED_ORB,
```
Can we get some comments or `#region`s about what these sections mean

Madmadness65 committed:
Remove temporary base sprites from ZA megas

Also added regions to the form change item enum file per request.

Madmadness65 committed:
Merge remote-tracking branch 'origin/release' into za-content

Madmadness65 committed:
Fix Greninja's forms after merge

Also updated Mega Garchomp Z's ability to Rough Skin, per request.

SirzBenjie committed:
Update `pokemon#getEffectiveStat`

Madmadness65 committed:
Fix failing Protosynthesis test

SirzBenjie committed:
Fix failing protosynthesis tests
